fdbserver/workloads/FuzzApiCorrectness.actor.cpp

```diff
@@ -749,6 +749,8 @@ struct FuzzApiCorrectnessWorkload : TestWorkload {
 				std::make_pair(error_code_tenant_not_found,
 				               ExceptionContract::possibleIf(!workload->canUseTenant(tr->getTenant()))),
 				std::make_pair(error_code_invalid_option,
+				               ExceptionContract::possibleIf(tr->getTenant().present() && specialKeys.contains(key))),
```

Comment:
It's possible invalid option is no longer possible now that we throw this new error. It may be worth trying to see.

Reply:
We still get invalid option errors in some cases, so I left it in

Reply:
Yes, it is when attempting to use RAW_ACCESS on a tenant transaction. That should be throwing `invalid_option`, correct?

Reply:
Is it using `RAW_ACCESS` by way of some special key-space module, or is the test setting `RAW_ACCESS`? If the former, then ideally we should be getting the other error (`illegal_tenant_access`), but if the latter then that's what we should expect. In that second case, though, we might be able to narrow the cases when we expect to see this error in the check here.

Reply:
https://github.com/apple/foundationdb/pull/7993/files#diff-74ee83047df93d7ae6bf63547f4e8b4b92742bb9f9d840539199f4967ee7a8bcR364-R393
It seems to be the former, and it happens before `SpecialKeySpace::getRangeAggregationActor` is able to check the ranges for the tenants. `normalizeKeySelectorActor`-> `moveKeySelectorOverRangeActor` -> `SpecialKeyRangeAsyncImpl::getRange`, which is overridden by several modules and then set RAW_ACCESS within those actors.  How come an actor meant to move a key selector ends up calling `getRange`? Doesn't that mean it will be called twice once we return to `SpecialKeySpace::getRangeAggregationActor`?
```
state RangeMap<Key, SpecialKeyRangeReadImpl*, KeyRangeRef>::Ranges ranges =
	    sks->getReadImpls().intersectingRanges(KeyRangeRef(begin.getKey(), end.getKey()));
```
We can't check tenant ranges until this is executed, and I don't think this can be executed before normalizing the key selectors, so the ordering here might not work out. Is the idea then to add the tenant's check to each module's `getRange` override if they set the RAW_ACCESS option?

Reply:
It looks like you may be right about calling get range twice on the same data there, but possibly that's because being more efficient is fairly complex? I'm not certain to be honest, but the code that does this kind of thing in RYW is certainly more complicated.

I suppose what we could do, then is to disallow even resolving the key selector over ranges that aren't supported by tenants. In other words, we could add a check in `moveKeySelectorOverRangeActor` (or `normalizeKeySelectorActor`) to verify that the range supports tenants or that we aren't using a tenant.

Reply:
Ok, I think that is a good solution. Would that mean also removing the check in `SpecialKeySpace::getRangeAggregationActor` or simply having it exist in both places at once?

Reply:
We probably need it in both. I haven't looked closely, but if I had to guess the key selector resolution may not read the entire range (say, if the begin was in one module and the end was in a later module, it may not read anything in between).